### ISSUE TITLE
LR scheduler design bug !

### ISSUE THREAD

**Issue by Kylin9511** [pytorch/pytorch#22107]
## 🐛 Bug

I really have to say: the new scheduler design in PyTorch1.1 is quite immature. Many bugs exist when using it. Related issue: #21623 (Although I think the usage in that issue is really wired.)

In my case, for MultiStepLR scheduler, get_lr function fails to demonstrate the correct learning rate when the step function is called. In fact, you will get a learning rate which is one more step further decayed. __In fact this problem exist for almost all the schedulers in torch1.1__

## To Reproduce

``` python
import torch
from torchvision.models import resnet18
net = resnet18()
optimizer = torch.optim.SGD(net.parameters(), 0.1)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)
for i in range(10):
    print(i, scheduler.get_lr())
    scheduler.step()
```

## Expected behavior

```
0 [0.1]
1 [0.1]
2 [0.1]
3 [0.0010000000000000002]
4 [0.010000000000000002]
5 [0.010000000000000002]
6 [0.00010000000000000003]
7 [0.0010000000000000002]
8 [0.0010000000000000002]
9 [1.0000000000000004e-05]
```

Appearently, the learning rate produced by scheduler.get_lr() when i in {3, 6, 9} is wrong. __10 times smaller than they should be, which is "one step further decayed"...__

## Environment

 - PyTorch Version : 1.1.0
 - Other environments does not matter. It is a code design bug. I will describe it briefly in the next part.

## analysis of the bug

From the source code, we can see the reason easily.

In `torch.optim.lr_scheduler._LRScheduler`, we can find the following function
``` python
def step(self, epoch=None):
    if epoch is None:
        epoch = self.last_epoch + 1
    self.last_epoch = epoch
    for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
        param_group['lr'] = lr
```

And in `torch.optim.lr_scheduler.MultiStepLR`, we can find the following function
``` python
def get_lr(self):
    if self.last_epoch not in self.milestones:
        return [group['lr'] for group in self.optimizer.param_groups]
    return [group['lr'] * self.gamma ** self.milestones[self.last_epoch]
            for group in self.optimizer.param_groups]
```

So let's say the lr need to be decayed on epoch 3. After the `step()` is called at the beginning of epoch 3, all the lr in `optimizer.param_groups` has already been decayed. And then we call `get_lr()` from the scheduler to get learning rate for log print, algorighm design or whatever. Ops.. the `get_lr()` function will return a decayed lr based on the already decayed lr  from `optimizer.param_groups`.

That is why under certain epochs, `get_lr()` function returns wrong lr which is __"one step further decayed"__

Of course, there exist one simple solution: just make sure the `scheduler.step()` is the last function you called during each epoch. This will fix the conflict but seems to be an ugly constraint. Also the design of scheduler actually demand to call `scheduler.step()` at the very beginning of each epoch.

So this solution is merely a emergency plan, and this is a design bug you need to fix. __My recommendation is to recover the self.base_lr in torch0.4.1.__

_p.s. There is a good news too: this bug does not influence vanilla training process. Merely wrong lr in log file._

**Comment by soumith** [comment]
@vincentqb @ezyang

**Comment by ezyang** [comment]
@luzhilin19951120 I think we fixed this on master; it was just an incorrect recursive formula for MultiStepLR. When I run your code I get:

```
(/scratch/ezyang/pytorch-build-env) ezyang@devfair040:/scratch/ezyang/pytorch-build$ python test.py
0 [0.1]
/scratch/ezyang/pytorch-build/torch/optim/lr_scheduler.py:82: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule.See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)
1 [0.1]
2 [0.1]
3 [0.010000000000000002]
4 [0.010000000000000002]
5 [0.010000000000000002]
6 [0.0010000000000000002]
7 [0.0010000000000000002]
8 [0.0010000000000000002]
9 [0.00010000000000000003]
```

Note I slightly simplified your example to get rid of the torchvision dependency:

```
import torch
from torch.nn import Parameter
optimizer = torch.optim.SGD([Parameter(torch.randn(2, 2, requires_grad=True))], 0.1)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)
for i in range(10):
    print(i, scheduler.get_lr())
    scheduler.step()
```

**Comment by ezyang** [comment]
Actually, never mind, on master we've only reverted the recursive formulas, #21800 isn't landed yet.

**Comment by ezyang** [comment]
So yes, I agree this is a legit problem, and we should not land #21800 until we figure out a way to solve this problem too.

**Comment by luzhilin19951120** [comment]
@ezyang Well, I see. You have already restored base_lr mechanism in master~
By the way, when do you plan to release PyTorch1.2.0 or 1.1.1 ?

**Comment by ezyang** [comment]
1.1.1 is unlikely to happen; 1.2.0 should happen in a few months.

Excerpts from Hubery Lu's message of 2019-06-24 22:11:44 -0700:
> @ezyang Well, I see. You have already restored base_lr mechanism in master~
> By the way, when do you plan to release PyTorch1.2.0 or 1.1.1 ?
>

**Comment by vincentqb** [comment]
As mentioned in #21800, the recursive behavior means that calling `get_lr` will compute the new learning rate based on the current learning rate (which has changed when using step). `get_lr` doesn't set the learning rate and so the training is not affected. The recursive code behave as it should, but this may be confusing as is happening here.

As mentioned [here](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485), `get_lr` is not a publicly documented function. We can move the computation logic away from `get_lr` to a private function, and make `get_lr` return the last computed learning rate by the current scheduler with a deprecation warning.

At the same time, we could also make `step` return its computed value. This would make it clear that the algorithm is also stepping forward. The current learning rate can also be access through `[p['lr'] for p in optimizer.param_groups]`.

Thoughts?

**Comment by luzhilin19951120** [comment]
@vincentqb I think keeping the old base_lr mechanisim will be better. 

The scheduler should be independent with the optimizer, do not make iterable call between them. After using optimizer to initialize the scheduler (and set the self.base_lr of scheduler of course), all the scheduler behaviour should not rely on the optimizer. 

__And I think you should add get_lr() to the document. It is a common request to get the current lr with a neat function during training.__

And the current function in MultiStep is really werid...
``` python
class MultiStepLR(_LRScheduler):
    def __init__(self, optimizer, milestones, gamma=0.1, last_epoch=-1):
        self.milestones = Counter(milestones)
        self.gamma = gamma
        super(MultiStepLR, self).__init__(optimizer, last_epoch)

    def get_lr(self):
        if self.last_epoch not in self.milestones:
            return [group['lr'] for group in self.optimizer.param_groups]
        return [group['lr'] * self.gamma ** self.milestones[self.last_epoch]
                for group in self.optimizer.param_groups]
```

__The implementation in the master branch makes me more comfortable...__

I think without strong reason, do not collete the design logic of schedulers. That may cause some unexpected trouble.

**Comment by luzhilin19951120** [comment]
@vincentqb Also, some extra scheduler may be added as well. like warm up, power function scheduler... But implementation of these scheduler is simple. Maybe your plan is to leave it to the user~ (I implement many other scheduler and a scheduler wrapper that can concatenate different scheduler. ^o^)

**Comment by vincentqb** [comment]
Thanks for the input!

> **And I think you should add get_lr() to the document. It is a common request to get the current lr with a neat function during training.**

Sounds good to me.

**Comment by vincentqb** [comment]
For reference, [here](https://github.com/pytorch/pytorch/pull/21800#issuecomment-506051090) is an explanation of what is happening in this pull request.

**Comment by vincentqb** [comment]
#24352

**Comment by Ir1d** [comment]
> I implement many other scheduler and a scheduler wrapper that can concatenate different schedulers
@Kylin9511 Hi Kylin, can you share your code for the wrapper? Thanks in advance.

**Comment by Kylin9511** [comment]
> > I implement many other scheduler and a scheduler wrapper that can concatenate different schedulers
> 
> @Kylin9511 Hi Kylin, can you share your code for the wrapper? Thanks in advance.

@Ir1d I afraid the implementation is highly merged in my pipeline. So the code will not provide much help. If you want to add the scheduler concatenation support into pytorch, you need to rewrite most of it.

**Comment by vincentqb** [comment]
#26423 has been merged and addresses this.

**Comment by VCBE123** [comment]
How to get the new feature? Now My Pytorch version is 1.10.

**Comment by vincentqb** [comment]
It's in master: you need to compile from source for now.

**Comment by neouyghur** [comment]
I find the same problem on 1.5 version.

**Comment by vincentqb** [comment]
Can you provide a minimal code snippet @neouyghur ? In particular, please make sure you are using `get_last_lr`, as mentioned in #26423.
